crowbar_framework/spec/models/api/upgrade_spec.rb

```diff
@@ -367,7 +379,175 @@
+      cinder_proposal.elements["cinder-controller"] = ["services"]
+      cinder_proposal.elements["cinder-volume"] = ["services"]
+      nova_proposal.elements["nova-controller"] = ["services"]
+      nova_proposal.elements["nova-compute-kvm"] = ["compute"]
+    end
+
+    it "leaves node with nova-compute role" do
+      proposals = {
+        "database" => database_proposal,
+        "cinder" => cinder_proposal,
+        "nova" => nova_proposal
+      }
+
+      expect(subject.class.upgradable_elements_of_proposals(proposals)).to(
+        eq(["data", "services"])
+      )
+    end
+
+    it "takes node with cinder-volume when alone" do
+      cinder_proposal.elements["cinder-volume"] = ["storage"]
+      proposals = {
+        "database" => database_proposal,
+        "cinder" => cinder_proposal,
+        "nova" => nova_proposal
+      }
+
+      expect(subject.class.upgradable_elements_of_proposals(proposals)).to(
+        eq(["data", "services", "storage"])
+      )
+    end
+
+    it "leaves node with cinder-volume when on nova-compute" do
+      cinder_proposal.elements["cinder-volume"] = ["compute"]
+      proposals = {
+        "database" => database_proposal,
+        "cinder" => cinder_proposal,
+        "nova" => nova_proposal
+      }
+
+      expect(subject.class.upgradable_elements_of_proposals(proposals)).to(
+        eq(["data", "services"])
+      )
+    end
+
+    it "takes node with cinder-volume when alone and in cluster" do
+      cinder_proposal.elements["cinder-volume"] = ["cluster:compute"]
+      proposals = {
+        "database" => database_proposal,
+        "cinder" => cinder_proposal,
+        "nova" => nova_proposal
+      }
+      allow(ServiceObject).to receive(:expand_nodes_for_all).and_return(
+        [["storage"]]
+      )
+
+      expect(subject.class.upgradable_elements_of_proposals(proposals)).to(
+        eq(["data", "services", "storage"])
+      )
+    end
+
+    it "leaves node with cinder-volume when in cluster with nova-compute" do
+      cinder_proposal.elements["cinder-volume"] = ["cluster:compute"]
+      proposals = {
+        "database" => database_proposal,
+        "cinder" => cinder_proposal,
+        "nova" => nova_proposal
+      }
+      allow(ServiceObject).to receive(:expand_nodes_for_all).and_return(
+        [["compute"]]
+      )
+
+      expect(subject.class.upgradable_elements_of_proposals(proposals)).to(
+        eq(["data", "services"])
+      )
+    end
+
+  end
+
+  context "upgrading the nodes in normal mode" do
+    it "successfully upgrades controller nodes" do
+      allow(Node).to(
+        receive(:find).with("state:crowbar_upgrade").
+        and_return([Node.find_by_name("testing.crowbar.com")])
+      )
+      allow(Api::Upgrade).to receive(:join_ceph_nodes).and_return(true)
+
+      allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:progress).and_return(
+        suggested_upgrade_mode: :normal
+      )
```

Comment:
you mean `allow_any_instance_of(Crowbar::UpgradeStatus).to receive(:upgrade_mode)` right?